### ISSUE TITLE
Segementation fault when calling a method

### ISSUE THREAD

**Issue by ulysseB** [rust-lang/rust#39292]
The following code generates a segfault when executed.
```rust
trait Foo<T> {
    fn print<'a>(&'a self) where T: 'a { println!("foo"); }
}

impl<'a> Foo<&'a ()> for () { }

trait Bar: for<'a> Foo<&'a ()> { }

impl Bar for () {}

fn main() {
    (&() as &Bar).print(); // Segfault
}
```

rust version:
```
rustc 1.16.0-nightly (9761b17d5 2017-01-21)
binary: rustc
commit-hash: 9761b17d558d0c980d4d2d5a0baba54f2bc6b63a
commit-date: 2017-01-21
host: x86_64-unknown-linux-gnu
release: 1.16.0-nightly
LLVM version: 3.9
````

**Comment by jonas-schievink** [comment]
Removing `where T: 'a` makes it work. The binary seems not to include the called method.

**Comment by jonas-schievink** [comment]
Suspicious code: https://github.com/rust-lang/rust/blob/1283c029557bcde106adde71f467fe133b3ffeff/src/librustc/traits/mod.rs#L619-L639 (seems to test whether `T: 'erased` holds to decide if the method can ever be called)

**Comment by nikomatsakis** [comment]
triage: P-high

Regression.

**Comment by arielb1** [comment]
This patch fixes the issue:
```
diff --git a/src/librustc/traits/mod.rs b/src/librustc/traits/mod.rs
index 4893e24091..71d67b106a 100644
--- a/src/librustc/traits/mod.rs
+++ b/src/librustc/traits/mod.rs
@@ -628,6 +628,7 @@ pub fn get_vtable_methods<'a, 'tcx>(
                                           |_, _| tcx.mk_region(ty::ReErased),
                                           |def, _| trait_ref.substs().type_for_def(def));
 
+            let substs = tcx.erase_late_bound_regions_and_normalize(&ty::Binder(substs));
             // It's possible that the method relies on where clauses that
             // do not hold for this particular set of type parameters.
             // Note that this method could then never be called, so we
```

**Comment by nikomatsakis** [comment]
Fix in #39887
